packages/by-name/kata/kata-runtime/0020-genpolicy-support-dynamic-annotations.patch

```diff
@@ -0,0 +1,177 @@
+---
+ src/tools/genpolicy/genpolicy-settings.json |  8 ++-
+ src/tools/genpolicy/rules.rego              | 55 +++++++++++++++++++--
+ src/tools/genpolicy/src/policy.rs           | 10 ++++
+ 3 files changed, 67 insertions(+), 6 deletions(-)
+
+diff --git a/src/tools/genpolicy/genpolicy-settings.json b/src/tools/genpolicy/genpolicy-settings.json
+index 9b95f9f7462717d04f0b9ce685d97c0455f949da..50406e8ac7a961d317f2254f30339c0c1d399f18 100644
+--- a/src/tools/genpolicy/genpolicy-settings.json
++++ b/src/tools/genpolicy/genpolicy-settings.json
+@@ -137,6 +137,9 @@
+             "io.katacontainers.pkg.oci.container_type": "pod_container",
+             "io.kubernetes.cri.container-type": "container"
+         },
++        "DynamicAnnotations": {
++            "^cdi\\.k8s\\.io\\/vfio[0-9]{2}$": "^nvidia.com/gpu=[0-9]+$"
++        },
+         "Linux": {
+             "Sysctl": {
+                 "net.ipv4.ip_unprivileged_port_start": "0",
+@@ -333,7 +336,8 @@
+                 "^AZURE_CLIENT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_TENANT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_FEDERATED_TOKEN_FILE=/var/run/secrets/azure/tokens/azure-identity-token$",
+-                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$"
++                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$",
++                "^PCI_RESOURCE_NVIDIA_COM.*=[a-fA-F0-9:.-]*$"
+             ]
+         },
+         "CopyFileRequest": [
+@@ -348,4 +352,4 @@
+         "UpdateEphemeralMountsRequest": false,
+         "WriteStreamRequest": false
+     }
+-}
+\ No newline at end of file
++}
+diff --git a/src/tools/genpolicy/rules.rego b/src/tools/genpolicy/rules.rego
+index 43cb19a56fe8ea5833708f0639c9e85ddd884cb3..554ab211b1392f5163cb894cfc29bc17cb4012d0 100644
+--- a/src/tools/genpolicy/rules.rego
++++ b/src/tools/genpolicy/rules.rego
+@@ -152,10 +152,10 @@ state_allows(key, value) = action {
+   state := get_state()
+   not state[key]
+   print("state_allows: saving to state key =", key, "value =", value)
+-  path := get_state_path(key) 
++  path := get_state_path(key)
+   action := {
+     "op": "add",
+-    "path": path, 
++    "path": path,
+     "value": value,
+   }
+ }
+@@ -199,13 +199,17 @@ allow_anno(p_oci, i_oci) {
+ }
+ allow_anno(p_oci, i_oci) {
+     print("allow_anno 2: p Annotations =", p_oci.Annotations)
++    print("allow_anno 2: p DynamicAnnotations =", p_oci.DynamicAnnotations)
++
++    i_annotations := i_oci.Annotations
+     print("allow_anno 2: i Annotations =", i_oci.Annotations)
+ 
+-    i_keys := object.keys(i_oci.Annotations)
++    i_keys := object.keys(i_annotations)
+     print("allow_anno 2: i keys =", i_keys)
+ 
+     every i_key in i_keys {
+         allow_anno_key(i_key, p_oci)
++        allow_dynamic_anno_value(i_key, i_annotations, p_oci)
+     }
+ 
+     print("allow_anno 2: true")
+@@ -227,6 +231,49 @@ allow_anno_key(i_key, p_oci) {
+     print("allow_anno_key 2: true")
+ }
+ 
++allow_anno_key(i_key, p_oci) {
```

Comment:
Did you consider adding the value to `allow_anno_key`? That way, you could write the check as

```rego
allow_anno(p_oci, i_oci) {
  # [...]
  every i_key, i_val in i_annotations {
    allow_anno_keyval(i_key, i_val, p_oci)
  }
}

allow_anno_keyval(i_key, i_val, p_oci) {
 # [...] Existing check
}

allow_anno_keyval(i_key, i_val, p_oci) {
  some p_key, p_val in p_oci.DynamicAnnotations
  regex.match(p_key, i_key)
  regex.match(p_val, i_val)
}
```

Reply:
I think this would make rebasing harder. What benefits do you see compared to the current implementation?

Reply:
The current implementation is probably wrong, looking at the negative check in `allow_dynamic_anno_value 2` which is only necessary because you're checking keys and values separately.

Reply:
Can you elaborate on that? What's wrong exactly?

Reply:
I also don't subscribe to the rebasing argument: you will have one line of diff in `allow_anno`, currently you have two. If you implement it as 

```rego
allow_anno_keyval(i_key, i_val) {
  allow_anno_key(i_key)
}

allow_anno_keyval(i_key, i_val) {
  # lookup in common.DynamicAnnotations ...
}
```

you don't need to touch any of the other existing predicates.

Reply:
> Can you elaborate on that? What's wrong exactly?

The list of dynamic annotations should be additive here, i.e. adding a new entry to dynamic annotations should never make my policy stricter. Suppose I added this to the settings:

```json
"dynamic_annotations": {"^.*$": "^$"}
```

Upstream currently allows all annotations that are explicitly added to the pod, but with this setting these would all be rejected because the key matches but the value doesn't.

Reply:
> Upstream currently allows all annotations that are explicitly added to the pod, but with this setting these would all be rejected because the key matches but the value doesn't.

But this would also be the case for your suggestion, wouldn't it?

Reply:
No, because the annotation would be accepted by the existing rule, which would be an alternative to the new rule.